### PR TITLE
FilterTable Developer Documentation

### PR DESCRIPTION
**Pull request by clintoncwolfe**
This PR adds two documents, intended for both the InSpec team and developers wishing to contribute.  It is not intended for usage by the general InSpec user community, and is not up to those standards.

Two documents are provided.

A usage document describes how a resource author would (and should) interact with FilterTable, and lists all known behaviors of FilterTable.  Extensive examples and gotchas are provided.

An internals document describes how FilterTable works on the inside.  FilterTable::Factory (the class that you use to configure your resource to work with FilterTable) and FilterTable::Table (the class that wraps the raw data and performs the filtration) are explored thoroughly.  Several areas for possible improvement are pointed out.

This PR reflects naming and line numbers prior to #3047 , so that the PRs may be considered independently.  The internals document in particular would need an update if that merges.

### CODE REVIEW

**Comment by aaronlippold** [comment]
On the test deployment I do not see links for this documentation or the other 'dev' docs as well. Where should I be looking?

**Comment by aaronlippold** [comment]
Many many thanks for a start on this - when this gets pushed in I am sure we will be pushing updates to this as well as the 'resource development' docs.

**Comment by clintoncwolfe** [comment]
That is correct, there are no links to this documentation, nor anything else in dev/docs at the moment.  This is an informal effort to capture tribal knowledge.  I imagine in the future we'll want to link or mention it in the Development section of the README; but I don't think this sort of material should be rendered and served on docs.inspec.io (at least not yet).

Long term, I think it would be great to have internal development documentation served on the docs website; but that introduces a higher standard of correctness, consistency, and thoroughness.  For the time being, I want it to be as easy as possible to "just take notes on what you learned, and don't worry about making it complete or nice"

File: docs/dev/filtertable-internals.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+The terminology used for many concepts does not help the reader understand what is going on.  Additionaly, the ways in which the classes relate is not straightforward.  Finally, variable names within the classes are often re-used ('filter' is a favorite) or are too short to be meaningful (x and c are both used as variable names, in long blocks).
+
+FilterTable was created in 2016 in an attempt to consolidate the pluralization features of several resources.  They each had slightly different featuresets, and were all in the wild, so FilterTable exposes some extensive side-effects to provide those features.
+
+## Where is the code?
+
+The main FilterTable code is in [utils/filter.rb](https://github.com/chef/inspec/blob/master/lib/utils/filter.rb).
+
+Also educational is the unit test for FIltertable, at test/unit/utils/filter_table_test.rb
```
Comment (jquick) [inline docs/dev/filtertable-internals.md]: FIltertable - think you need lowercase `I`

File: docs/dev/filtertable-internals.md [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+```
+things.smaller_than(12)
+```
+
+If you provide _both_ a block and opts, only the block is used, and the options are ignored.
+
+#### Behavior when no block is provided
+
+If you do not provide a block, you _must_ provide a `:field` option (though that does no appear to be enforced). The behavior is to define a method with the name provided, that has a conditional return type. The methodd body is defined in lines 258-266.
```
Comment (jquick) [inline docs/dev/filtertable-internals.md]: *Methodd*

File: docs/dev/filtertable-internals.md [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+#### Re-pack the "connectors"
+
+First, on lines 188-192, the list of custom methods ("connectors", registered using the `add` method) are repacked into an array of arrays of two elements - the desired method name and the lambda that will be used as the method body.  The lambda is created by the private method `create_connector`.
+
+TBD: what exactly create_connector does
+
+#### Defines a special Struct type to represent rows in the table
+
+At lines 195-200, a new Struct type is defined, with attributes for each of the known table fields.  THe motivation fot this struct type is to implement the block-mode behavior of `where`.  Because each struct represents a row, and it has the attributes (accessors) for the fields, block-mode `where` is implemented by instance-evaling against each row as a struct.
```
Comment (jquick) [inline docs/dev/filtertable-internals.md]: *THe*, *fot*

File: docs/dev/filtertable-internals.md [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+Line 224 uses `send` with a block to call `define_method` on the resource class that you're authoring.  Using a block with `send` is undocumented, but is treated as an implicit argument (per stackoverflow) , so the end result is that the block is used as the body for the new method being defined.
+
+The method body is wrapped in an exception-catching facility that catches skipped or failed resource excepytions and wraps them in a specialized excepytion catcher class. TBD: understand this better.
+
+Line 226 constructs an instance of the anonymous FilterTable::Table subclass defined at 203.  It passes three args:
+
+1. `self`. TBD: which class is this referring to at this point?
+2. The return value of calling the data fetcher method.
+3. The string ' with', which is probably informing the criteria stringification?
```
Comment (jquick) [inline docs/dev/filtertable-internals.md]: *' with'* (extra space)

File: docs/dev/filtertable-internals.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+And that new FilterTable::Table subclass instance is stored in a local variable, named, confusingly, "filter".
+
+On line 227, we then immediately call a method on that "FilterTable::Table subclass instance". The method name is the same as the one we're defining on the resource - but we're calling it on the Table.  Recall we defined all the "connector" methods on the Table subclass at line 204-206.  The method gets called with any args or block passed, and since it's the last thing, it provides the return value.
+
+VERY WORRISOME THING: So, the Table subclass has methods for the "connectors" (for example, `thing_ids` or `exist?`.  What about the "accessors" - `where` and `entries`?  Are those in the FilterTable::Table class, or method_missing'd?
```
Comment (jquick) [inline docs/dev/filtertable-internals.md:214]: *method_missing'd*

**Review by jquick — APPROVED** [review]
Thanks for this @clintoncwolfe, just a few typo's

File: docs/dev/filtertable-internals.md (hunk 4 above)
Comment (clintoncwolfe) [reply docs/dev/filtertable-internals.md]: The extra space is correct - I'll add clarification

**Pull request merged**